core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala

```diff
@@ -199,6 +200,18 @@ private[spark] class HeartbeatReceiver(sc: SparkContext, clock: Clock)
       if (now - lastSeenMs > executorTimeoutMs) {
         logWarning(s"Removing executor $executorId with no recent heartbeats: " +
           s"${now - lastSeenMs} ms exceeds timeout $executorTimeoutMs ms")
+        sc.schedulerBackend match {
+          case backend: CoarseGrainedSchedulerBackend =>
+            backend.synchronized {
+              // Mark executor pending to remove if executor heartbeat expired
+              // to avoid reschedule task on this executor again
+              if (!backend.executorsPendingToRemove.contains(executorId)) {
+                backend.executorsPendingToRemove(executorId) = false
```

Comment:
I think this does not resolve the issue completely. Because when you look into `sc.killAndReplaceExecutor` below, you can see that we've already mark it as "pending to remove". Though it happens in another separate thread, but I don't see big difference according to this issue. WDYT?

Reply:
> `sc.killAndReplaceExecutor` already try to mark it as "pending to remove" this is right, but the task has rescheduled at this executor agait at this time, the executor must be removed from the ExecutorBackend to avoid, for example, it will `disableExecutor` in `CoarseGrainedSchedulerBackend` if the driver lost connection from executor.

Reply:
The task can reschedule at this executor before mark it as "pending to remove".

Reply:
We add the executor before rescheduler can avoid the tasks to rescheduler the bad executors

Reply:
Hi, @seayoun. After a second think, I think this fix can really avoid app hang, though it can not clean up dead records in `CoarseGrainedSchedulerBackend`. 

As you may have realized that a same issue has been posted in SPARK-27348 and its author(@xuanyuanking) has assigned to me recently. And it's really coincidental. And I really appreciate that you would like my proposal. Actually, the main idea is still base on original author's contribution.

Also, thanks for review.